test/e2e_node/restart_test.go

```diff
@@ -0,0 +1,117 @@
+			continue
+		}
+
+		runningPods = []*api.Pod{}
+		for _, pod := range podList.Items {
+			if r, err := testutils.PodRunningReady(&pod); err != nil {
+				continue
+			} else if !r {
+				continue
+			}
+			runningPods = append(runningPods, &pod)
+		}
+		framework.Logf("Running pod count %d", len(runningPods))
+		if len(runningPods) >= pod_count {
+			break
+		}
+	}
+	return runningPods
+}
+
+var _ = framework.KubeDescribe("Restart [Serial] [Slow] [Disruptive]", func() {
+	const (
+		// Saturate the node. It's not necessary that all these pods enter
+		// Running/Ready, because we don't know the number of cores in the
+		// test node or default limits applied (if any). It's is essential
+		// that no containers end up in terminated. 100 was chosen because
+		// it's the max pods per node.
+		POD_COUNT             = 100
+		POD_CREATION_INTERVAL = 100 * time.Millisecond
+		RECOVER_TIMEOUT       = 5 * time.Minute
+		START_TIMEOUT         = 3 * time.Minute
+		MIN_PODS              = 20
+		RESTART_COUNT         = 6
+	)
+
+	f := framework.NewDefaultFramework("restart-test")
+	Context("Docker Daemon", func() {
+		Context("Network", func() {
+			It("should recover from ip leak", func() {
+
+				pods := newTestPods(POD_COUNT, framework.GetPauseImageNameForHostArch(), "restart-docker-test")
+				By(fmt.Sprintf("Trying to create %d pods on node", len(pods)))
+				createBatchPodWithRateControl(f, pods, POD_CREATION_INTERVAL)
+				defer deletePodsSync(f, pods)
+
+				// Give the node some time to stabilize, assume pods that enter RunningReady within
+				// START_TIMEOUT fit on the node and the node is now saturated.
+				runningPods := waitForPods(f, POD_COUNT, START_TIMEOUT)
+				if len(runningPods) < MIN_PODS {
+					framework.Failf("Failed to start %d pods, cannot test that restarting docker doesn't leak IPs", MIN_PODS)
+				}
+
+				for i := 0; i < RESTART_COUNT; i += 1 {
+					By(fmt.Sprintf("Restarting Docker Daemon iteration %d", i))
+
+					// TODO: Find a uniform way to deal with systemctl/initctl/service operations. #34494
+					if stdout, err := exec.Command("sudo", "systemctl", "restart", "docker").CombinedOutput(); err != nil {
+						framework.Logf("Failed to trigger docker restart with systemd/systemctl: %v, stdout: %q", err, string(stdout))
+						if stdout, err = exec.Command("sudo", "service", "docker", "restart").CombinedOutput(); err != nil {
+							framework.Failf("Failed to trigger docker restart with upstart/service: %v, stdout: %q", err, string(stdout))
+						}
+					}
+					time.Sleep(20 * time.Second)
+				}
+
+				By("Checking currently Running/Ready pods")
+				postRestartRunningPods := waitForPods(f, len(runningPods), RECOVER_TIMEOUT)
+				if len(postRestartRunningPods) == 0 {
+					framework.Failf("Failed to start *any* pods after docker restart, this might indicate an IP leak")
+				}
+				By("Confirm no containers have terminated")
+				for _, pod := range postRestartRunningPods {
```

Comment:
yeah that's lastState, im checking state which shouldn't be terminated at the end of this experiment (it is in step 3 below):    

1. schedule new pods on a full ip node, they end up with state waiting (containerCreating, no available ips). 
2. If a pod gets a chance to run, we have state running (startedAt blah). 
3. If docker gets bounced and we end up with no available ips, we have state terminated (finishedAt blah), Ready=false.
4. Now GC runs and frees up some ips (in fact with 100 pods gc keeps running after the 3rd restart).
5. And we get state running (startedAt blah) with lastState terminated (finishedAt blah), Ready=true.

If GC hadn't run, we would be stuck at (3) a 100 containers with state terminated (finishedAt blah), ready=false.